src/simulators/density_matrix/densitymatrix_thrust.hpp

```diff
@@ -454,84 +459,217 @@ double DensityMatrixThrust<data_t>::probability(const uint_t outcome) const {
+#endif
+
+  BaseVector::UpdateReferencedValue();
+
   samples.assign(SHOTS, 0);
 
-  const int INDEX_SIZE = BaseVector::sample_measure_index_size_;
-  const int_t INDEX_END = BITS[INDEX_SIZE];
-  // Qubit number is below index size, loop over shots
-  if (END < INDEX_END) {
-    #pragma omp parallel if (BaseVector::num_qubits_ > BaseVector::omp_threshold_ && BaseVector::omp_threads_ > 1) num_threads(BaseVector::omp_threads_)
-    {
-      #pragma omp for
-      for (int_t i = 0; i < SHOTS; ++i) {
-        double rnd = rnds[i];
-        double p = .0;
-        int_t sample;
-        for (sample = 0; sample < END - 1; ++sample) {
-          p += probability(sample);
-          if (rnd < p)
-            break;
-        }
-        samples[i] = sample;
+  localSamples.assign(SHOTS, 0);
+
+  placeSum.assign(BaseVector::m_nPlaces+1, 0.0);
+
+  //diagonal indices for each place
+  for(iPlace=0;iPlace<BaseVector::m_nPlaces;iPlace++){
+    i = BaseVector::m_Chunks[iPlace].GlobalIndex();
+    diagIdx[iPlace] = i >> num_qubits();
+    if(i > (diagIdx[iPlace] << num_qubits())){
+      diagIdx[iPlace]++;
+    }
+  }
+  i = BaseVector::m_Chunks[BaseVector::m_nPlaces-1].GlobalIndex() + BaseVector::m_Chunks[BaseVector::m_nPlaces-1].Size();
+  diagIdx[BaseVector::m_nPlaces] = i >> num_qubits();
+
+  //calculate sum of each place
+#pragma omp parallel if (BaseVector::num_qubits_ > BaseVector::omp_threshold_ && BaseVector::m_nPlaces > 1) num_threads(BaseVector::m_nPlaces) private(pVec,iPlace)
+  {
+    int iDev;
+    uint_t nDiag;
+
+    iPlace = omp_get_thread_num();
+    iDev = BaseVector::m_Chunks[iPlace].DeviceID();
+#ifdef AER_THRUST_CUDA
+    if(iDev >= 0){
+      cudaSetDevice(iDev);
+    }
+#endif
+
+    nDiag = diagIdx[iPlace+1]-diagIdx[iPlace];
+    if(nDiag > 0){
+      uint_t pos = diagIdx[iPlace] * ((1ull << num_qubits()) + 1) - BaseVector::m_Chunks[iPlace].GlobalIndex();
+
+      pVec = (data_t*)(BaseVector::m_Chunks[iPlace].ChunkPtr(0,BaseVector::m_maxChunkBits) + pos);
+
+      auto ci = thrust::counting_iterator<uint_t>(0);
+
+      //gather diagonal elements, then inclusive scan them
+      if(iPlace < BaseVector::m_nDevParallel){
+        thrust::for_each_n(thrust::device,ci,nDiag,DensityMatrixThrust_Gather<data_t>(pVec,(1ull << num_qubits())));
+
+        thrust::inclusive_scan(thrust::device,pVec,pVec+nDiag,pVec);
       }
-    } // end omp parallel
+      else{
+        auto policy = (BaseVector::num_qubits_ > BaseVector::omp_threshold_ && BaseVector::omp_threads_ > 1)
+            ? thrust::omp::par
+            : thrust::seq;
+        thrust::for_each_n(policy,ci,nDiag,DensityMatrixThrust_Gather<data_t>(pVec,(1ull << num_qubits())));
+
+        thrust::inclusive_scan(policy,pVec,pVec+nDiag,pVec);
+      }
+
+      thrust::complex<data_t> v = BaseVector::m_Chunks[iPlace].GetState(pos + (nDiag-1)/2);
+      if((nDiag & 1) == 0)
+        placeSum[iPlace] = v.imag();
+      else
+        placeSum[iPlace] = v.real();
+    }
+    else{
```

Comment:
Let's remove this `else` by moving the sentence below to whatever `if` condition belongs, like:
```c++
placeSum[iPlace] = 0;
if( ... ) {

}
```